src/lang/checkers/SanityChecker.ml

```diff
@@ -69,12 +77,104 @@ module ScillaSanityChecker
+        (* Walk the syntax tree looking for code paths containing one
+           or more accept statements.
+
+           seen is a List with an item for each code path which has
+           already seen at least one accept statement before reaching
+           the point in the syntax tree just before stmts.  Each item
+           in the List is List of all the accepts seen along that path
+           to this point.
+
+           Returns an updated version of the List of Lists which
+           includes any accepts seen in stmts.
+
+           Note that any given code path remains untracked until an
+           accept is discovered in it.  This keeps the seen data
+           nested List to a minimal size. *)
+        List.fold_left
+          (fun (seen2 : loc list list) (stmt : stmt_annot) ->
+            let loc = (stmt_loc stmt) in
+            match fst stmt with
+            | AcceptPayment ->
+               (match seen2 with
+                (* If we didn't see any accept statements before reaching
+                 * this point, register this as the first one. *)
+                | [] -> [[loc]]
+                (* Otherwise add this accept statement to the list of accepts
+                 * already seen on each code path reaching this point. *)
+                | _ -> List.map (fun accepts -> loc :: accepts) seen2)
+            | MatchStmt (_ident, branches) ->
+               (* For each branch in the match statement we have a
+                    new code path to "multiply" with the code paths
+                    which already reached this point, so walk each
+                    branch and build all the results into a new list.
+                *)
+               List.fold_left
+                 (fun seen3 (_pattern, branchstmts) ->
+                   match walk seen2 branchstmts with
+                   | [] -> seen3
+                   | seen4 -> seen3 @ seen4)
+                 [] branches
+            | _ -> seen2
+          ) (seen : loc list list) (stmts : stmt_annot list)
+      in
+        walk [] stmts
+    in
+
+    let check_accepts (contr : contract) =
+      let check_transition_accepts (transition : transition) =
+        let transition_accept_groups = List.map List.rev (find_accept_groups transition.tbody) in
+
+        let dup_accept_warning (group : loc list) : unit =
+          (warn1
+             (Core.sprintf "transition %s had a potential code path with duplicate accept statements:\n"
+                (get_id transition.tname) ^
+                String.concat ""
+                  (List.map
+                     (fun loc -> Core.sprintf "  Accept at %s\n" (get_loc_str loc))
+                     group))
+             warning_level_duplicate_accepts
+             (List.hd group))
+        in
+
+        List.iter
+          (fun group ->
+            match List.length group with
+            | 0 -> ()
+            | 1 -> ()
+            | _ -> dup_accept_warning group)
```

Comment:
`List.length` traverses the entire list, so it takes time proportional to the length of the list. Since this check is equivalent to testing whether the length is greater than 1, you can instead do the following:

```
match group with
| _ :: _ :: _ -> dup_accept_warning group
| _ -> ()
```

Or you can introduce a helper function `list_empty : 'a List -> bool`.

Reply:
Fixed via `match`.  BTW I notice that `batteries` provides `BatList.is_empty` so presumably that would have been another option?

Reply:
Batlist.is_empty does the same thing. I am just not used to using Batteries.